### PR TITLE
create_vms & sushy | Allow disabling Secure Boot mode

### PR DESCRIPTION
**Pull request by tkrishtop**
- This is required for some use cases, e.g. run tests with syscall iopl()
- Allowing to disable Secure Boot mode with a flag variable, false by default
- This affects to create_vms and setup_sushy_tools roles. Some variables needed to be updated to follow collections' best practices.
- Changed VM creation script, not using `--boot uefi` mode, just to be able to reuse the variables we're defining for the OVMF path. The proposed config for the "enabled secure boot" case is exactly what would be generated when using `--boot uefi` mode

Test-Hints: no-check

build-depends: 31929

- [x] ABI job with secure boot enabled - https://www.distributed-ci.io/jobs/c7f0114c-bb1c-44c3-a8b8-3c7cd166f775/jobStates?sort=date

```
$ sudo virsh dumpxml dciokd-worker-1
...
  <os>                                                      
    <type arch='x86_64' machine='pc-q35-rhel8.6.0'>hvm</type>                    
    <loader readonly='yes' secure='yes' type='pflash'>/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd</loader>
    <nvram template='/usr/share/edk2/ovmf/OVMF_VARS.secboot.fd'>/var/lib/libvirt/qemu/nvram/dciokd-worker-1_VARS.fd</nvram>
    <boot dev='fd'/>
  </os>
  <features>
    <acpi/>
    <apic/>
    <smm state='on'/>
  </features>
...

$ ssh core@dciokd-worker-1.server06.partnerci.bos2.lab
...
---
[core@dciokd-worker-1 ~]$ dmesg | grep secure
[    0.000000] secureboot: Secure boot enabled
[    0.012221] secureboot: Secure boot enabled
```

- [x] ABI job with secure boot disabled - https://www.distributed-ci.io/jobs/a7155ff1-2f16-4889-9707-341b4ad4fddc/jobStates?sort=date

```
$ sudo virsh dumpxml dciokd-worker-1
...
  <os>
    <type arch='x86_64' machine='pc-q35-rhel8.6.0'>hvm</type>
    <loader readonly='yes' secure='no' type='pflash'>/usr/share/OVMF/OVMF_CODE.secboot.fd</loader>
    <nvram>/var/lib/libvirt/qemu/nvram/dciokd-worker-1_VARS.fd</nvram>
    <boot dev='fd'/>
  </os>
  <features>
    <acpi/>
    <apic/>
  </features>
...

$ ssh core@dciokd-worker-1.server03.partnerci.bos2.lab
...
---
[core@dciokd-worker-1 ~]$ dmesg | grep secure
[    0.000000] secureboot: Secure boot disabled
[    0.012947] secureboot: Secure boot disabled
```

### CODE REVIEW

**Comment by nicomola** [comment]
from change https://github.com/redhatci/ansible-collection-redhatci-ocp/pull/367:
- FAILURE https://www.distributed-ci.io/jobs/ee9929db-9c8b-4503-ad02-07743faad69c/jobStates

**Comment by ramperher** [comment]
recheck

**Comment by ramperher** [comment]
I can see that, now, the XML domain description is correct:

```
$ sudo virsh dumpxml dciokd-master-0
...
  <os>
    <type arch='x86_64' machine='pc-q35-rhel8.6.0'>hvm</type>
    <loader readonly='yes' secure='no' type='pflash'>/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd</loader>
    <nvram>/var/lib/libvirt/qemu/nvram/dciokd-master-0_no_secure_boot_VARS.fd</nvram>
    <boot dev='fd'/>
  </os>
...
```

Also, `/var/lib/libvirt/qemu/nvram/dciokd-master-0_no_secure_boot_VARS.fd` is created correctly (same for other VMs):

```
$ ls -la /var/lib/libvirt/qemu/nvram/
total 3176
drwxr-xr-x.  2 qemu qemu   4096 Jul  4 10:29 .
drwxr-x--x. 26 qemu qemu   4096 Jul  4 10:28 ..
-rw-r--r--.  1 root root 540672 Jul  4 10:28 dciokd-master-0_no_secure_boot_VARS.fd
-rw-r--r--.  1 root root 540672 Jul  4 10:28 dciokd-master-1_no_secure_boot_VARS.fd
-rw-r--r--.  1 root root 540672 Jul  4 10:28 dciokd-master-2_no_secure_boot_VARS.fd
-rw-r--r--.  1 root root 540672 Jul  4 10:28 dciokd-worker-0_no_secure_boot_VARS.fd
-rw-r--r--.  1 root root 540672 Jul  4 10:28 dciokd-worker-1_no_secure_boot_VARS.fd
-rw-r--r--.  1 root root 540672 Jul  4 10:29 dciokd-worker-2_no_secure_boot_VARS.fd
```

However, job is failing in the same step it was failing previously with the change in sushy-tools:

```
    \"message\": \"Unknown boot loader path \\\"/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd\\\" in domain \\\"fc749263-ef08-421c-9833-c01c20f3cf4f\\\" configuration encountered while setting boot mode \\\"UEFI\\\", system architecture \\\"x86_64\\\". Consider adding this loader path to emulator config.\",
```

This needs to be checked.

File: roles/create_vms/templates/create_vm.sh.j2
Comment (manurodriguez) [inline roles/create_vms/templates/create_vm.sh.j2:29]: >   \"message\": \"Unknown boot loader path \\\"/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd\\\" in domain \\\"fc749263-ef08-421c-9833-c01c20f3cf4f\\\" configuration encountered while setting boot mode \\\"UEFI\\\", system architecture \\\"x86_64\\\". Consider adding this loader path to emulator config.\",

I see the Hypervisors default loader points to `/usr/share/OVMF/OVMF_CODE.secboot.fd`

```
[dci@server09 ~]$ sudo virsh domcapabilities              
<domainCapabilities>                           
  <path>/usr/libexec/qemu-kvm</path>                  
  <domain>kvm</domain>                                          
  <machine>pc-i440fx-rhel7.6.0</machine>            
  <arch>x86_64</arch>                             
  <vcpu max='240'/>                              
  <iothreads supported='yes'/>                   
  <os supported='yes'>                           
    <enum name='firmware'/>                                
    <loader supported='yes'>                   
      <value>/usr/share/OVMF/OVMF_CODE.secboot.fd</value>
      <enum name='type'>                         
        <value>rom</value>                  
        <value>pflash</value>               
      </enum>                                             
      <enum name='readonly'>                       
        <value>yes</value>                            
        <value>no</value>                        
      </enum>                                  
      <enum name='secure'>                
        <value>no</value>              
      </enum>                             
    </loader>                             
  </os>  
...  
```
And that path is a symbolic link to `/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd` 
```
[dci@server09 ~]$ ls -l /usr/share/OVMF/OVMF_CODE.secboot.fd
lrwxrwxrwx. 1 root root 33 Mar 14 08:56 /usr/share/OVMF/OVMF_CODE.secboot.fd -> ../edk2/ovmf/OVMF_CODE.secboot.fd
```
So it is the same file, but in your xml config the loader points to `/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd` so I wonder if you change it to match the same path as shown in the `virsh domcapabilities` output, that is to use the path `/usr/share/OVMF/OVMF_CODE.secboot.fd` everywhere to be consistent. 

I manually powered a VM from the last job failure, and it booted, so I got the impression all looks good in libvirt, it's only sushy that complains right?, where did you get that error @ramperher ? 

Initially I had some concerns that OVMF_CODE.secboot.fd only works with secure boot enabled, as Andreas described it in this blog https://andreaskaris.github.io/blog/linux/libvirt-uefi-without-secureboot/ where he even uses an non-secure boot UEFI firmware, but it might be an old bug.

Comment (manurodriguez) [reply roles/create_vms/templates/create_vm.sh.j2:29]: I see now where did you get the error, when KVM is trying to Set OneTimeBoot Virtual Media. I went to the logs carefully, so it's really sushy that complains : )

I think I managed to pass that error, long story short, I tested my theory above and if you change `--boot loader=/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd` to `--boot loader=/usr/share/OVMF/OVMF_CODE.secboot.fd` in the VM definition, then we are able to set the OneTimeBoot.

I replicated the issue with a basic play, using dciokd-master-1 in server09:
```
---
- name: Play to validate stuff
  hosts: localhost
  gather_facts: false
  vars:
    system_uuid_url: "https://server09-assisted.partnerci.bos2.lab:8082/redfish/v1/Systems/fc749263-ef08-421c-9833-c01c20f3cf4f"                                                                                                             
    bmc_user: USER
    bmc_password: PASSWORD
    target_host: dciokd-master-1
  tasks:

    - name: KVM Set OneTimeBoot Virtual Media {{ target_host }}
      uri:
        url: "{{ system_uuid_url }}"
        user: "{{ bmc_user }}"
        password: "{{ bmc_password }}"
        method: PATCH
        body_format: json
        body:
          {
            "Boot":
              {
                "BootSourceOverrideTarget": "Cd",
                "BootSourceOverrideMode": "UEFI",
                "BootSourceOverrideEnabled": "Continuous",
              },
          }
        status_code: [200, 204]
        force_basic_auth: false
        validate_certs: false
        return_content: true

[dci@server09 ~]$ sudo virsh dumpxml dciokd-master-1 | grep uuid
  <uuid>fc749263-ef08-421c-9833-c01c20f3cf4f</uuid>

```

Then I tried to set the OneTimeBoot when the VM had a loader pointing to `/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd`  and it fails.
```
[dci@server09 ~]$ sudo virsh dumpxml dciokd-master-1 | grep loader
    <loader readonly='yes' secure='no' type='pflash'>/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd</loader>

[dci-openshift-agent@server01 manny]$ ansible-playbook simple-play.yml                                                                                                                                                                       
PLAY [Play to validate stuff] ****************************************************************************************************************************************************************************************************************

TASK [KVM Set OneTimeBoot Virtual Media dciokd-master-1] *************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "content": "{\n  \"error\": {\n    \"code\": \"Base.1.0.GeneralError\",\n    \"message\": \"Unknown boot loader path \\\"/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd\\\" in domain \\\"fc749263-ef08-421c-9833-c01c20f3cf4f\\\" configuration encountered while setting boot mode \\\"UEFI\\\", system architecture \\\"x86_64\\\". Consider adding this loader path to emulator config.\",\n    \"@Message.ExtendedInfo\": [\n      {\n        \"@odata.type\": \"/redfish/v1/$metadata#Message.1.0.0.Message\",\n        \"MessageId\": \"Base.1.0.GeneralError\"\n      }\n    ]\n  }\n}", "content_length": "510", "content_type": "application/json", "date": "Thu, 04 Jul 2024 23:30:54 GMT", "elapsed": 0, "json": {"error": {"@Message.ExtendedInfo": [{"@odata.type": "/redfish/v1/$metadata#Message.1.0.0.Message", "MessageId": "Base.1.0.GeneralError"}], "code": "Base.1.0.GeneralError", "message": "Unknown boot loader path \"/usr/share/edk2/ovmf/OVMF_CODE.secboot.fd\" in domain \"fc749263-ef08-421c-9833-c01c20f3cf4f\" configuration encountered while setting boot mode \"UEFI\", system architecture \"x86_64\". Consider adding this loader path to emulator config."}}, "msg": "Status code was 500 and not [200, 204]: HTTP Error 500: INTERNAL SERVER ERROR", "redirected": false, "server": "Werkzeug/2.0.3 Python/3.6.8", "status": 500, "url": "https://server09-assisted.partnerci.bos2.lab:8082/redfish/v1/Systems/fc749263-ef08-421c-9833-c01c20f3cf4f"}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

Then, I changed the loader to point to /usr/share/OVMF/OVMF_CODE.secboot.fd, and this time the play passed
```
[dci@server09 ~]$ sudo virsh edit dciokd-master-1
Domain 'dciokd-master-1' XML configuration edited.

[dci@server09 ~]$ sudo virsh dumpxml dciokd-master-1 | grep loader
    <loader readonly='yes' secure='no' type='pflash'>/usr/share/OVMF/OVMF_CODE.secboot.fd</loader>

[dci-openshift-agent@server01 manny]$ ansible-playbook simple-play.yml                                                                                                                                                                       
PLAY [Play to validate stuff] ****************************************************************************************************************************************************************************************************************

TASK [KVM Set OneTimeBoot Virtual Media dciokd-master-1] *************************************************************************************************************************************************************************************
ok: [localhost]

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

Then this makes me believe that in the VM definition we need to use the same loader path as defined in sushy config:
```
[dci@server09 ~]$ grep -A4 SUSHY_EMULATOR_BOOT_LOADER_MAP /opt/dci/assisted/sushy/sushy-emulator.conf 
SUSHY_EMULATOR_BOOT_LOADER_MAP = {
    u'UEFI': {
        u'x86_64': u'/usr/share/OVMF/OVMF_CODE.secboot.fd',
        u'aarch64': u'/usr/share/AAVMF/AAVMF_CODE.fd'
    },
```

Comment (ramperher) [reply roles/create_vms/templates/create_vm.sh.j2]: Good observation @manurodriguez , thanks for checking this :) I'll change the boot loader to use `/usr/share/OVMF/OVMF_CODE.secboot.fd` now and let's see.

Comment (ramperher) [reply roles/create_vms/templates/create_vm.sh.j2]: Now KVM failed task is now passing, I think this can be closed, let's see if the job passes.

**Comment by ramperher** [comment]
Secure boot looks like disabled now:

```
$ ssh core@dciokd-worker-1.server10.partnerci.bos2.lab
Warning: Permanently added 'dciokd-worker-1.server10.partnerci.bos2.lab,192.168.16.105' (ECDSA) to the list of known hosts.
Red Hat Enterprise Linux CoreOS 416.94.202407030122-0
  Part of OpenShift 4.16, RHCOS is a Kubernetes-native operating system
  managed by the Machine Config Operator (`clusteroperator/machine-config`).

WARNING: Direct SSH access to machines is not recommended; instead,
make configuration changes via `machineconfig` objects:
  https://docs.openshift.com/container-platform/4.16/architecture/architecture-rhcos.html

---
[core@dciokd-worker-1 ~]$ dmesg | grep Secure
[    0.000000] secureboot: Secure boot disabled
[    0.011880] secureboot: Secure boot disabled
[    1.949104] integrity: Loaded X.509 cert 'Red Hat Secure Boot CA 5: cc6fa5e72868ba494e939bbd680b9144769a9f8f'
[    1.951930] integrity: Loaded X.509 cert 'Red Hat Secure Boot CA 8: e1c6c580aa1e21d585aad9bf20f3929e5ec1f08b'

```

Waiting for the job to pass, this change would be ready in that case.

**Review by ramperher — DISMISSED** [review]
LGTM, job is passing and the setup is behaving as expected.

**Review by manurodriguez — APPROVED** [review]
LGTM :tada:

**Comment by ramperher** [comment]
@nocturnalastro , I think we need a review from a crucible team member before merging, is this change fine from your side?

**Comment by ramperher** [comment]
recheck

**Comment by nicomola** [comment]
from change https://github.com/dci-labs/keysight-lab-config/pull/217:
- SUCCESS https://www.distributed-ci.io/jobs/6da779f7-b87e-4ca2-8812-93e62f381455/jobStates

**Comment by ramperher** [comment]
Adding also @thekad since he's also included in crucible project, just to see if the review can be unblocked.

File: roles/create_vms/templates/create_vm.sh.j2
Comment (nocturnalastro) [inline roles/create_vms/templates/create_vm.sh.j2:5]: Is there a need to copy? Can't you just point the loader at the original?

File: roles/setup_sushy_tools/tasks/main.yml
Comment (nocturnalastro) [inline roles/setup_sushy_tools/tasks/main.yml:87]: This shouldn't be changed, the path you've changed it to gets stat-ed in the next task and replaced if its found.

**Review by nocturnalastro — CHANGES_REQUESTED** [review]

File: roles/setup_sushy_tools/tasks/main.yml
Comment (nocturnalastro) [reply roles/setup_sushy_tools/tasks/main.yml:87]: If we need the same path for sushy and the vm then we should duplicate/point the logic we have here in the VM role.

Comment (ramperher) [reply roles/setup_sushy_tools/tasks/main.yml:87]: I'll retry with removing this change, to see how it goes

File: roles/create_vms/templates/create_vm.sh.j2
Comment (ramperher) [reply roles/create_vms/templates/create_vm.sh.j2:5]: Yes, because we need to generate one OVMF_VARS.fd file per node (copy contains `item.name`), which is later on used as argument for the `--boot` option, with `nvram=/var/lib/libvirt/qemu/nvram/{{ item.name }}_no_secure_boot_VARS.fd`. I've just done that because I saw that, by default, each VM uses a different VARS.fd file, which is created automatically.
So, I think it's fine to leave this in this way.

File: roles/setup_sushy_tools/tasks/main.yml
Comment (ramperher) [reply roles/setup_sushy_tools/tasks/main.yml]: We definitely need this change in the OVMF code path, because it's failing when trying to boot the VM: https://www.distributed-ci.io/jobs/9c30d6e3-6dc3-4aa8-b0c9-ed34b1ebd1ec/jobStates?sort=date&task=f9f27ff9-1f3d-4218-a5a2-b6e6d37f4bbd, so I'll add a check for disabling secure boot in the sushy tools role too

Comment (ramperher) [reply roles/setup_sushy_tools/tasks/main.yml]: Now it's passing, with the new refactor of setup_sushy_tools: https://www.distributed-ci.io/jobs/ecb35cf2-d87e-4df4-ad1a-cc1f99f78f8f/jobStates?sort=date&task=43930d9d-a4c9-4209-9cb5-de850b000741

File: roles/create_vms/templates/create_vm.sh.j2
Comment (thekad) [inline roles/create_vms/templates/create_vm.sh.j2:3]: this comment should be inside the if block

File: roles/create_vms/templates/create_vm.sh.j2
Comment (thekad) [inline roles/create_vms/templates/create_vm.sh.j2:5]: consider having all these OVMF paths as variables

**Review by thekad — APPROVED** [review]
minor nitpicks for a follow-up PR, all good

**Comment by ramperher** [comment]
> minor nitpicks for a follow-up PR, all good

Thanks @thekad for your review! I'll check it here because I need to do a refactor after talking with @nocturnalastro , so I'll apply your changes here.

File: roles/create_vms/templates/create_vm.sh.j2
Comment (ramperher) [reply roles/create_vms/templates/create_vm.sh.j2:3]: ack

File: roles/create_vms/templates/create_vm.sh.j2
Comment (ramperher) [reply roles/create_vms/templates/create_vm.sh.j2:5]: ack

File: roles/create_vms/defaults/main.yml
Comment (tonyskapunk) [inline roles/create_vms/defaults/main.yml:1]: Since this new functionality affects the DCI Agent, it needs a bump in the collection version and the RPM to satisfy its dependencies.

Comment (ramperher) [reply roles/create_vms/defaults/main.yml:1]: ack

**Comment by ramperher** [comment]
Change is ready for review again, I've included in the PR description the successful results for both enabled/disabled secure boot cases with the corresponding output of the interesting VM info to confirm that they're working as expected. Jobs are about to pass.

File: roles/create_vms/tasks/provision_vms.yml
Comment (nocturnalastro) [inline roles/create_vms/tasks/provision_vms.yml:45]: I don't see why you need this override here and is sushy tools

If both VMs and sushy tools is using the same logic they should always align so you don't need to force it.

**Review by nocturnalastro — APPROVED** [review]
We've agreed to simplify the logic here soon but for now this will suffice.

File: roles/create_vms/tasks/provision_vms.yml
Comment (ramperher) [reply roles/create_vms/tasks/provision_vms.yml:45]: Discussed offline, also with @p3ck , we'll check the refactor of the OVMF path in a separate change, just to have this feature already integrated in the collections.

**Comment by nicomola** [comment]
from change https://github.com/dci-labs/keysight-lab-config/pull/217:
- SUCCESS https://www.distributed-ci.io/jobs/f1e35ee8-0f76-40d8-8f02-a7daf26628b5/jobStates

**Pull request merged**
